costmap_2d/include/costmap_2d/footprint.h

```diff
@@ -40,18 +40,106 @@
+ */
 geometry_msgs::Point32            toPoint32    (geometry_msgs::Point   pt);
+
+/**
+ * @brief Convert vector of Points to Polygon msg
+ */
 geometry_msgs::Polygon            toPolygon    (std::vector<geometry_msgs::Point> pts);
+
+/**
+ * @brief Convert Polygon msg to vector of Points. 
+ */
 std::vector<geometry_msgs::Point> toPointVector(geometry_msgs::Polygon polygon);
 
+/**
+ * @brief  Given a pose and base footprint, build the oriented footprint of the robot (list of Points)
+ * @param  x The x position of the robot
+ * @param  y The y position of the robot
+ * @param  theta The orientation of the robot
+ * @param  footprint_spec Basic shape of the footprint
+ * @param  oriented_footprint Will be filled with the points in the oriented footprint of the robot
+*/
+void transformFootprint(double x, double y, double theta, const std::vector<geometry_msgs::Point>& footprint_spec, std::vector<geometry_msgs::Point>& oriented_footprint);
+
+/**
+ * @brief  Given a pose and base footprint, build the oriented footprint of the robot (PolygonStamped)
+ * @param  x The x position of the robot
+ * @param  y The y position of the robot
+ * @param  theta The orientation of the robot
+ * @param  footprint_spec Basic shape of the footprint
+ * @param  oriented_footprint Will be filled with the points in the oriented footprint of the robot
+*/
+void transformFootprint(double x, double y, double theta, const std::vector<geometry_msgs::Point>& footprint_spec, geometry_msgs::PolygonStamped & oriented_footprint);
+
+/**
+ * @brief Adds the specified amount of padding to the footprint (in place)
+ */
+void padFootprint( std::vector<geometry_msgs::Point>& footprint, double padding );
+
+/**
+ * @brief Create a circular footprint from a given radius
+ */
+std::vector<geometry_msgs::Point> makeFootprintFromRadius( double radius );
+
+/** 
+ * @brief Make the footprint from the given string.
+ *
+ * Format should be bracketed array of arrays of floats, like so: [[1.0, 2.2], [3.3, 4.2], ...] 
+ * 
+ * If there's an error, the error_flag will be set to true. 
+ */
+std::vector<geometry_msgs::Point> makeFootprintFromString( const std::string& footprint_string, bool* error_flag );
+
+/** 
+ * @brief Read the ros-params "footprint" and/or "robot_radius" from
+ * the given NodeHandle using searchParam() to go up the tree.
+ */
+std::vector<geometry_msgs::Point> makeFootprintFromParams( ros::NodeHandle& nh );
+
+/** 
+ * @brief Create the footprint from the given XmlRpcValue.
+ *
+ * @param footprint_xmlrpc should be an array of arrays, where the
+ * top-level array should have 3 or more elements, and the
+ * sub-arrays should all have exactly 2 elements (x and y
+ * coordinates).
+ *
+ * @param full_param_name this is the full name of the rosparam from
+ * which the footprint_xmlrpc value came.  It is used only for
+ * reporting errors. */
+std::vector<geometry_msgs::Point> makeFootprintFromXMLRPC( XmlRpc::XmlRpcValue& footprint_xmlrpc,
```

Comment:
For some reason involving the ROS debug macros, adding const (`makeFootprintFromXMLRPC( const XmlRpc::XmlRpcValue& footprint_xmlrpc,`) results in a number of weird errors.